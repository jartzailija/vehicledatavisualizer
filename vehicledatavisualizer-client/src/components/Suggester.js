import React from 'react';
import Autosuggest from 'react-autosuggest';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Suggester = ({name, description, keywordSuggestion, suggestions, changeText, select}) => {
  //styled-components doesn't work with react-autosuggest
  const theme = {

    input: {
      width: '410px'
    },

    suggestionsContainer: {
      position: 'absolute',
      backgroundColor: 'white',
      zIndex: 100
    },
    suggestionsList: {
      listStyleType: 'none',
      border: '1px solid black',
      padding: '3px',
      width: '410px',
      backgroundColor: 'white',
      marginBottom: 0
    }
  };

	const getSuggestionValue = suggestion => suggestion;

	const renderSuggestion = suggestion => {
		return (
      <div className="suggestion">
        {suggestion}
      </div>
    );
  };

	const getSuggestions = value => {
		const inputValue = value.trim().toLowerCase();
		const inputLength = inputValue.length;
		return inputLength === 0 ? [] : suggestions.filter(sug =>
			sug.toLowerCase().slice(0, inputLength) === inputValue
		);
	};
	
	const onSuggestionsFetchRequested = ({ value }) => {
    suggestions = getSuggestions(value);
    return suggestions;
	};

	const onSuggestionsClearRequested = () => {
    suggestions = [];
    return suggestions;
	};

	const inputProps = {
		placeholder: description,
		value: keywordSuggestion,
		onChange: changeText
	};

	return (
    <div className="wrapper">
      <Autosuggest
        theme={theme}
        suggestions={suggestions}
        onSuggestionsFetchRequested={onSuggestionsFetchRequested}
        onSuggestionsClearRequested={onSuggestionsClearRequested}
        onSuggestionSelected={select}
        getSuggestionValue={getSuggestionValue}
        renderSuggestion={renderSuggestion}
        alwaysRenderSuggestions={true}
        inputProps={inputProps}
      />
    </div>
	);
}

Suggester.propTypes = {
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  keywordSuggestion: PropTypes.string.isRequired,
  suggestions: PropTypes.array.isRequired,
  selectedKeyword: PropTypes.string.isRequired,
  changeText: PropTypes.func.isRequired,
  select: PropTypes.func.isRequired
};

export default Suggester;