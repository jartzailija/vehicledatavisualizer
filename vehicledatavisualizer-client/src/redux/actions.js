import configs from '../configs';

export const SET_INPUT = 'SET_INPUT';
export const setInput = municipality =>  {
  return { type: SET_INPUT, payload: municipality };
};

export const SELECT_MUNICIPALITY = 'SELECT_MUNICIPALITY';
export const selectMunicipality = municipality =>  {
  return { type: SELECT_MUNICIPALITY, payload: municipality };
};

export const UNSELECT_MUNICIPALITY = 'UNSELECT_MUNICIPALITY';
export const unselectMunicipality = () =>  {
  return { type: UNSELECT_MUNICIPALITY };
};

export const SET_SUGGESTIONS = 'SET_SUGGESTIONS';
export const setSuggestions = suggestions =>  {
  return { type: SET_SUGGESTIONS, payload: suggestions };
};

export const EMPTY_SUGGESTIONS = 'EMPTY_SUGGESTIONS';
export const emptySuggestions = () =>  {
  return { type: EMPTY_SUGGESTIONS };
};

export const HAS_CARBRANDS = 'HAS_CARBRANDS';
export const setHavingCarBrands = yesOrNo =>  {
  return { type: HAS_CARBRANDS, payload: yesOrNo };
};

export const SET_CARBRANDS = 'SET_CARBRANDS';
export const setCarBrands = data =>  {
  return { type: SET_CARBRANDS, payload: data };
};

export const RESET_CARBRANDS = 'RESET_CARBRANDS';
export const resetCarBrands = () =>  {
  return { type: RESET_CARBRANDS };
};

export const SET_CARBRANDS_COUNT = 'SET_CARBRANDS_COUNT';
export const setCarBrandsCount = count =>  {
  return { type: SET_CARBRANDS_COUNT, payload: count };
};

export const SELECT_CARBRAND = 'SELECT_CARBRAND';
export const selectCarBrand = carBrand =>  {
  return { type: SELECT_CARBRAND, payload: carBrand };
};

export const UNSELECT_CARBRAND = 'UNSELECT_CARBRAND';
export const unSelectCarBrand = () =>  {
  return { type: UNSELECT_CARBRAND };
};

export const SET_AVERAGE_DATA = 'SET_AVERAGE_DATA';
export const setAverageData = dataObj =>  {
  return { type: SET_AVERAGE_DATA, payload: { ...dataObj }};
};

export const EMPTY_AVERAGE_DATA = 'EMPTY_AVERAGE_DATA';
export const emptyAverageData = () =>  {
  return { type: EMPTY_AVERAGE_DATA };
};

export const fetchSuggestions = (name, keyword) => {
  const editedKeyword = keyword.charAt(0).toUpperCase() + keyword.substring(1);
  const lowName = name.toLowerCase();

  return dispatch => {
    //TODO: siirr채 konffifiluun
    return fetch(`${configs.frontEndUrl}/api/values/${lowName}/${editedKeyword}`)
      .then(response => {
        return response.json();
      })
      .then(json => {
        const suggestions = json.map(item => item.key);
        dispatch(setSuggestions(suggestions));
      })
      .catch(e => console.log(e));
  };
};

export const fetchDataByKeyword = (name, keyword) => {
  const editedKeyword = keyword.charAt(0).toUpperCase() + keyword.substring(1);
  const lowName = name.toLowerCase();

  return dispatch => {
    dispatch(emptyAverageData());
    dispatch(unSelectCarBrand());
    dispatch(resetCarBrands());
    //TODO: siirr채 konffifiluun
    return fetch(`${configs.frontEndUrl}/api/query/${lowName}/${editedKeyword}`)
      .then(response => {
        return response.json();
      })
      .then(json => {
        dispatch(setHavingCarBrands(true));
        dispatch(setCarBrands(json.carBrands));
        dispatch(setCarBrandsCount(json.count));
      })
      .catch(e => console.log(e));
  }
};

export const getDataFromCarBrand = (brandName, municipality) => {

  const start = `${configs.frontEndUrl}/api/query/summary/`
  const end = municipality === '' ? `carbrand/${brandName}` : `municipality/${municipality}/carbrand/${brandName}`;
  return dispatch => {
    dispatch(emptyAverageData());
    dispatch(selectCarBrand(brandName));
    //TODO: siirr채 konffifiluun
    return fetch(start + end)
      .then(response => response.json())
      .then(json => dispatch(setAverageData(json)))
      .catch(e => console.log(e));
  }
};

export const getCountrywideData = () => {
  return dispatch => {
    dispatch(emptyAverageData());
    dispatch(unSelectCarBrand());
    dispatch(resetCarBrands());
    dispatch(unselectMunicipality());
    //TODO: siirr채 konffifiluun
    return fetch(`${configs.frontEndUrl}/api/query/all`)
      .then(response => response.json())
      .then(json => {
        dispatch(setHavingCarBrands(true));
        dispatch(setCarBrands(json.carBrands));
        dispatch(setCarBrandsCount(json.count));
      })
      .catch(e => console.log(e));
  }
};