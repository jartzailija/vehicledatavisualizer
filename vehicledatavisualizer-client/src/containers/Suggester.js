import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Suggester from '../components/Suggester';
import {
  setInput,
  selectMunicipality,
  fetchSuggestions,
  fetchDataByKeyword,
  emptyAverageData,
  emptySuggestions} from '../redux/actions';

const mapStateToProps = (state, ownProps) => {
  return {
    ...ownProps,
    keywordSuggestion: state.municipalities.input,
    suggestions: state.municipalities.suggestions,
    selectedKeyword: state.municipalities.selected
  };
};

const mapDispatchToProps = (dispatch, ownProps) => ({
  changeText: (event, { newValue, method }) => {
    dispatch(setInput(newValue));
    if(newValue.length >= 3) {
      dispatch(fetchSuggestions(ownProps.name, newValue));
    }
  },
  select: (event, { suggestion, suggestionValue, suggestionIndex, sectionIndex, method }) => {
    dispatch(selectMunicipality(suggestion));
    dispatch(emptySuggestions(ownProps.name));
    dispatch(emptyAverageData());
    dispatch(fetchDataByKeyword(ownProps.name, suggestion));
  }
});

const ConnectSuggester = connect(
  mapStateToProps,
  mapDispatchToProps
)(Suggester);

ConnectSuggester.propTypes = {
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired
};

export default ConnectSuggester;