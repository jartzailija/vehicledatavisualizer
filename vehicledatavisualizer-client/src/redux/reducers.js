import { combineReducers } from 'redux';
import {
  SET_INPUT,
  SELECT_MUNICIPALITY,
  UNSELECT_MUNICIPALITY,
  SET_SUGGESTIONS,
  EMPTY_SUGGESTIONS,
  HAS_CARBRANDS,
  SET_CARBRANDS,
  SET_CARBRANDS_COUNT,
  SELECT_CARBRAND,
  SET_AVERAGE_DATA,
  EMPTY_AVERAGE_DATA,
  RESET_CARBRANDS,
  UNSELECT_CARBRAND
} from './actions';


const initialMunicipalityState = {
	selected: '',
	input: '',
	suggestions: [],
};

const updateMunicipality = (state = initialMunicipalityState, action) => {
  switch (action.type) {
    case SET_INPUT:
      return {
        ...state,
        input: action.payload
      };
    
    case SET_SUGGESTIONS:
      return {
        ...state,
        suggestions: action.payload
      };

    case EMPTY_SUGGESTIONS:
      return {
        ...state,
        suggestions: []
      };

    case SELECT_MUNICIPALITY:
      return {
        ...state,
        selected: action.payload
      };

    case UNSELECT_MUNICIPALITY:
      return {
        ...state,
        selected: ''
      };
      
    default:
      return { ...state };
  }
};

const initialCarBrandsState = {
  hasCarBrands: false,
  data: [],
  count: 0,
  selected: ''
};

const carBrands = (state = initialCarBrandsState, action) => {
  switch(action.type) {
    case HAS_CARBRANDS:
      return {
        ...state,
        hasCarBrands: action.payload
      };

    case SET_CARBRANDS:
      return {
        ...state,
        data: action.payload
      };

    case SET_CARBRANDS_COUNT:
      return {
        ...state,
        count: action.payload
      };

    case SELECT_CARBRAND:
      return {
        ...state,
        selected: action.payload
      };

    case UNSELECT_CARBRAND:
      return {
        ...state,
        selected: ''
      };

    case RESET_CARBRANDS:
      return { ...initialCarBrandsState };

    default:
      return { ...state };
  }
};

const initialDataState = {
  hasData: false,
  age: 0,
  mileage: 0,
  count: 0,
  motorSize: 0,
  power: 0
};

const averageData = (state = initialDataState, action) => {
  switch(action.type) {
    case SET_AVERAGE_DATA:
      return {
        ...state,
        hasData: true,
        age: action.payload.age,
        mileage: action.payload.mileage,
        count: action.payload.count,
        motorSize: action.payload.motorSize,
        power: action.payload.power
      };

    case EMPTY_AVERAGE_DATA:
      return { ...initialDataState };

    default:
      return { ...state };
  };
};

const vehicleVisualizer = combineReducers({
  municipalities: updateMunicipality,
  carBrands,
  averageData
});

export default vehicleVisualizer;
