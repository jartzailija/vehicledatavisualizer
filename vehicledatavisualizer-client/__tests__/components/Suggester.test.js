import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { createMockStore } from "redux-test-utils";
import shallowWithStore from '../utils/shallowWithStore';
import {setInput} from '../../src/redux/actions';

// Component to be tested
import Suggester from '../../src/components/Suggester';
import SuggesterContainer, {mapDispatchToProps, mapStateToProps} from '../../src/containers/Suggester';

describe('<Suggester />', () => {

  const mockState = {
    municipalities: {
      input: 'kunta',
      suggestions: ['kunta1', 'kunta2'],
      selected: ''
    }
  };
  let store;
  //let container; 

  beforeEach(() => {
    store = createMockStore(mockState);
  });

  it('renders', () => {
    const container = shallowWithStore(<SuggesterContainer name="name" description="description" />, store);
    expect(container).toBeTruthy();
  });

  //No need to test connection, it's taken care by the 3rd party library
  //TODO: JATKA
  it('test dispatch', () => {
    const dispatch = jest.fn();

    const readyDispatch = mapDispatchToProps(dispatch, {name: 'j'})
    readyDispatch.changeText('', {newValue: 'jo', method: ''});
    expect(dispatch.mock.calls[0][0]).toEqual({ type: 'SET_INPUT', payload: 'jo'});

    readyDispatch.select();
  });

  it('gets the state', () => {

  });
});