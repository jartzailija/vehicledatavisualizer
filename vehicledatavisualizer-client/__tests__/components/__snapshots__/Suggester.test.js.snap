// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Suggester /> the proper component renders the component with data 1`] = `
<div
  className="wrapper"
>
  <Autosuggest
    alwaysRenderSuggestions={true}
    focusInputOnSuggestionClick={true}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "onChange": [Function],
        "placeholder": "description",
        "value": "",
      }
    }
    multiSection={false}
    onSuggestionSelected={[Function]}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={Array []}
    theme={
      Object {
        "input": Object {
          "width": "410px",
        },
        "suggestionsContainer": Object {
          "backgroundColor": "white",
          "position": "absolute",
          "zIndex": 100,
        },
        "suggestionsList": Object {
          "backgroundColor": "white",
          "border": "1px solid black",
          "listStyleType": "none",
          "marginBottom": 0,
          "padding": "3px",
          "width": "410px",
        },
      }
    }
  />
</div>
`;
